## Importando bibliotecas

In [1]:
import pandas as pd

## Definindo funções úteis
Em todo o script que crio costumo definir algumas funções para alterar a tipagem de colunas:

In [2]:
## converter a coluna para string
def to_str(df, coluna):
    df[coluna] = df[coluna].astype('str')

## converter a coluna para int
def to_int(df, coluna):
    df[coluna] = df[coluna].astype('int')

## converter a coluna para float
def to_float(df, coluna):
    df[coluna] = df[coluna].astype('float')

## converter a coluna para datetime
def to_datetime(df, coluna):
    df[coluna] = pd.to_datetime(df[coluna])

## Base de clientes

Lendo a base:

In [3]:
df_clientes = pd.read_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Primarias/base_clientes.xlsx")

Primeiramente, irei verificar as colunas dessa base:

In [4]:
df_clientes.columns

Index(['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], dtype='object')

Agora, utilizando o .head(), irei fazer uma breve análise de como está estruturada essa base de dados:

In [5]:
df_clientes.head()

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,Base de clientes atualizado até 31/02/2028,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN
4,CPF,Nome,Cidade,Estado,CEP


In [6]:
df_clientes = df_clientes.dropna()

In [7]:
df_clientes.columns = ['cpf', 'nome', 'cidade', 'estado', 'cep']

In [8]:
df_clientes = df_clientes[df_clientes['cpf'] != 'CPF']

Verificando se a alteração foi feita com sucesso:

In [9]:
df_clientes.head()

,cpf,nome,cidade,estado,cep
5,54747820977,Larissa Carvalho,Palmas,TO,77000-000
6,05887824075,Juliana Martins,Palmas,TO,77000-000
7,07289624387,Larissa Souza,Salvador,BA,40000-000
8,00112020761,Lucas Souza,Paranaguá,PR,83200-000
9,00831518129,Juliana Carvalho,Ponta Grossa,PR,84000-000


Agora irei verificar se todos os valores na coluna CPF possuem a mesma quantidade de caracteres:

In [10]:
to_str(df_clientes, 'cpf')
df_clientes['cpf'].str.len().nunique()

1

O retorno foi 3, o que significa que sim, existem valores que não possuem a mesma quantidade de caracteres. Por padrão, CPF's possuem 11 caracteres, neste momento não sei se isso irá impactar em futuros relacionamentos ou análises, portanto irei manter dessa forma e caso seja necessário alterarei posteriormente.

A seguir, verificarei demais informações na base, a fim de identificar possíveis missings (dados nulos):

In [11]:
df_clientes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 999 entries, 5 to 1003
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cpf     999 non-null    object
 1   nome    999 non-null    object
 2   cidade  999 non-null    object
 3   estado  999 non-null    object
 4   cep     999 non-null    object
dtypes: object(5)
memory usage: 46.8+ KB


Visto que não há valores nulos, agora vou verificar se há algum CPF duplicado, o que seria incorreto visto que essa é uma base dimensão:

In [12]:
df_clientes['cpf'].duplicated().any()

False

Visto que não há duplicidades, irei baixar a base tratada:

In [13]:
df_clientes.to_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Secundarias/tb_clientes.xlsx", index=False)

## Base de chamadas

Lendo a base:

Na primeira tentativa de ler o arquivo recebi uma mensagem de erro de codificação, para solucionar esse problema irei especificar uma codificação diferente de UTF-8:

In [14]:
df_chamadas = pd.read_csv("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Primarias/base_chamadas.csv", encoding='ISO-8859-1')

Após carregar a base, vou verificar as colunas:

In [15]:
df_chamadas.columns

Index(['data;;;;;hora;retido_ura;recebido;atendido;;;tempo_espera;tempo_atendimento;telefone;id_operador;nome_operador;tabulação;subtabulacao'], dtype='object')

Visto que o separador de colunas dessa base é ";", irei ler novamente o arquivo com o separador ";":

In [16]:
df_chamadas = pd.read_csv("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Primarias/base_chamadas.csv", encoding='ISO-8859-1', sep=';')

Verificando novamente as colunas:

In [17]:
df_chamadas.columns

Index(['data', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'hora',
       'retido_ura', 'recebido', 'atendido', 'Unnamed: 9', 'Unnamed: 10',
       'tempo_espera', 'tempo_atendimento', 'telefone', 'id_operador',
       'nome_operador', 'tabulação', 'subtabulacao'],
      dtype='object')

Foi possível verificar que existem diversas colunas sem nome específicado, vou verificar se existem dados nelas para descobrir se serão utilizadas ou não:

In [18]:
df_chamadas.head()

,data,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,hora,retido_ura,recebido,atendido,Unnamed: 9,Unnamed: 10,tempo_espera,tempo_atendimento,telefone,id_operador,nome_operador,tabulação,subtabulacao
0,29/01/2025,NaN,NaN,NaN,NaN,21:29:18,0,1,0,NaN,NaN,00:00:52,00:03:53,79193777494,142012,Luna Ferreira da Rocha,4. Canal de Atendimento,4.2 E-mail
1,07/01/2025,NaN,NaN,NaN,NaN,00:16:27,0,1,1,NaN,NaN,00:00:45,00:08:34,27803480622,1752,Levi Cunha da Costa,2. Status do Atendimento,2.4 Em Análise
2,05/01/2025,NaN,NaN,NaN,NaN,06:46:34,1,0,0,NaN,NaN,00:00:24,00:03:31,23796752326,31913,Helena Costa Ferreira,6. Ação Tomada,6.4 Devolução/Reembolso
3,24/01/2025,NaN,NaN,NaN,NaN,14:57:04,0,1,1,NaN,NaN,00:01:06,00:07:09,85385553021,11204,João Costela Viana,4. Canal de Atendimento,4.2 E-mail
4,27/01/2025,NaN,NaN,NaN,NaN,23:55:41,0,1,1,NaN,NaN,00:01:43,00:03:44,33181611619,2134,Ian Pires Peixoto,5. Grau de Satisfação,5.1 Muito Satisfeito


In [19]:
df_chamadas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9998 entries, 0 to 9997
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   data               9998 non-null   object 
 1   Unnamed: 1         0 non-null      float64
 2   Unnamed: 2         0 non-null      float64
 3   Unnamed: 3         0 non-null      float64
 4   Unnamed: 4         0 non-null      float64
 5   hora               9998 non-null   object 
 6   retido_ura         9998 non-null   int64  
 7   recebido           9998 non-null   int64  
 8   atendido           9998 non-null   int64  
 9   Unnamed: 9         0 non-null      float64
 10  Unnamed: 10        0 non-null      float64
 11  tempo_espera       9998 non-null   object 
 12  tempo_atendimento  9998 non-null   object 
 13  telefone           9998 non-null   object 
 14  id_operador        9998 non-null   int64  
 15  nome_operador      9998 non-null   object 
 16  tabulação          9998 

Visto que de fato as colunas sem nome não possuem dados, irei selecionar aquelas que serão realmente necessárias:

In [20]:
df_chamadas = df_chamadas[['data', 'hora', 'retido_ura', 'recebido', 'atendido', 'tempo_espera', 'tempo_atendimento', 'telefone', 'id_operador', 'nome_operador', 'tabulação', 'subtabulacao']]

Verificando a alteração:

In [21]:
df_chamadas.head()

,data,hora,retido_ura,recebido,atendido,tempo_espera,tempo_atendimento,telefone,id_operador,nome_operador,tabulação,subtabulacao
0,29/01/2025,21:29:18,0,1,0,00:00:52,00:03:53,79193777494,142012,Luna Ferreira da Rocha,4. Canal de Atendimento,4.2 E-mail
1,07/01/2025,00:16:27,0,1,1,00:00:45,00:08:34,27803480622,1752,Levi Cunha da Costa,2. Status do Atendimento,2.4 Em Análise
2,05/01/2025,06:46:34,1,0,0,00:00:24,00:03:31,23796752326,31913,Helena Costa Ferreira,6. Ação Tomada,6.4 Devolução/Reembolso
3,24/01/2025,14:57:04,0,1,1,00:01:06,00:07:09,85385553021,11204,João Costela Viana,4. Canal de Atendimento,4.2 E-mail
4,27/01/2025,23:55:41,0,1,1,00:01:43,00:03:44,33181611619,2134,Ian Pires Peixoto,5. Grau de Satisfação,5.1 Muito Satisfeito


Agora irei verificar se há missings (valores nulos):

In [22]:
df_chamadas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9998 entries, 0 to 9997
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   data               9998 non-null   object
 1   hora               9998 non-null   object
 2   retido_ura         9998 non-null   int64 
 3   recebido           9998 non-null   int64 
 4   atendido           9998 non-null   int64 
 5   tempo_espera       9998 non-null   object
 6   tempo_atendimento  9998 non-null   object
 7   telefone           9998 non-null   object
 8   id_operador        9998 non-null   int64 
 9   nome_operador      9998 non-null   object
 10  tabulação          9998 non-null   object
 11  subtabulacao       9998 non-null   object
dtypes: int64(4), object(8)
memory usage: 937.4+ KB


Visto que não há, vou verificar a tipagem das colunas, tendo em vista que há colunas de data:

In [23]:
df_chamadas.dtypes

data                 object
hora                 object
retido_ura            int64
recebido              int64
atendido              int64
tempo_espera         object
tempo_atendimento    object
telefone             object
id_operador           int64
nome_operador        object
tabulação            object
subtabulacao         object
dtype: object

In [24]:
to_datetime(df_chamadas, 'data')
df_chamadas['hora'] = pd.to_datetime(df_chamadas['hora'], format='%H:%M:%S').dt.time
df_chamadas['tempo_espera'] = pd.to_datetime(df_chamadas['tempo_espera'], format='%H:%M:%S').dt.time
df_chamadas['tempo_atendimento'] = pd.to_datetime(df_chamadas['tempo_atendimento'], format='%H:%M:%S').dt.time

C:\Users\Igor\AppData\Local\Temp\ipykernel_25180\3171289282.py:15: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[coluna] = pd.to_datetime(df[coluna])


Por hora irei manter essas alterações, caso necessário farei outras posteriormente.

Verificando as modificações:

In [25]:
df_chamadas.head()

,data,hora,retido_ura,recebido,atendido,tempo_espera,tempo_atendimento,telefone,id_operador,nome_operador,tabulação,subtabulacao
0,2025-01-29,21:29:18,0,1,0,00:00:52,00:03:53,79193777494,142012,Luna Ferreira da Rocha,4. Canal de Atendimento,4.2 E-mail
1,2025-01-07,00:16:27,0,1,1,00:00:45,00:08:34,27803480622,1752,Levi Cunha da Costa,2. Status do Atendimento,2.4 Em Análise
2,2025-01-05,06:46:34,1,0,0,00:00:24,00:03:31,23796752326,31913,Helena Costa Ferreira,6. Ação Tomada,6.4 Devolução/Reembolso
3,2025-01-24,14:57:04,0,1,1,00:01:06,00:07:09,85385553021,11204,João Costela Viana,4. Canal de Atendimento,4.2 E-mail
4,2025-01-27,23:55:41,0,1,1,00:01:43,00:03:44,33181611619,2134,Ian Pires Peixoto,5. Grau de Satisfação,5.1 Muito Satisfeito


Padronizando o nome das colunas em minúsculo e sem pontuações.

Verificando os nomes atuais:

In [26]:
df_chamadas.columns

Index(['data', 'hora', 'retido_ura', 'recebido', 'atendido', 'tempo_espera',
       'tempo_atendimento', 'telefone', 'id_operador', 'nome_operador',
       'tabulação', 'subtabulacao'],
      dtype='object')

Neste caso apenas irei remover as pontuações da coluna 'tabulação':

In [27]:
df_chamadas = df_chamadas.rename(columns={'tabulação' : 'tabulacao'})

Validando a alteração:

In [28]:
df_chamadas.columns

Index(['data', 'hora', 'retido_ura', 'recebido', 'atendido', 'tempo_espera',
       'tempo_atendimento', 'telefone', 'id_operador', 'nome_operador',
       'tabulacao', 'subtabulacao'],
      dtype='object')

Para facilitar futuros relacionamentos, irei alterar a tipagem de algumas colunas:

In [29]:
to_str(df_chamadas, 'id_operador')
to_str(df_chamadas, 'telefone')

Baixando a base tratada:

In [30]:
df_chamadas.to_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Secundarias/tb_chamadas.xlsx", index=False)

## Base operadores

Lendo a base

In [31]:
df_operadores = pd.read_fwf("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Primarias/base_operadores.txt")

Verificando as colunas

In [32]:
df_operadores.columns

Index(['id_operador', '0 - 15', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], dtype='object')

Verificando a estrutura da base

In [33]:
df_operadores.head(8)

,id_operador,0 - 15,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,nome_operador,16 - 40,NaN,NaN,NaN
1,equipe,56 - 15,NaN,NaN,NaN
2,supervisor,71 - 40,NaN,NaN,NaN
3,carga_horária,111 - 15,NaN,NaN,NaN
4,1132,Enrico Duarte Nascimento,Platina,Nina Mendes Gonçalves,7h12
5,3119,Isabella Moraes Gonçalves,Diamante,Isabel da Mota da Luz,7h12
6,81711,Mariana Gonçalves Aragão,Ouro,Larissa Nunes Pereira,7h12
7,622,Marcela Novaes Freitas,Bronze,Larissa Nunes Pereira,7h12


Aparentemente as colunas estão dispostas da seguinte maneira, em ordem: id_operador, nome_operador, equipe, supervisor, carga_horária.
Em uma situação real de trabalho eu procuraria o responsável pela base ou outra pessoa que pudesse me auxiliar em entender o que os valores de "0-15", "16-40", etc.. significam. Neste caso, irei supor que não são relevantes nesse momento e irei ignorar eles. 

Isto posto, agora irei atribuir o nome das colunas corretamente.

In [34]:
# Primeiramente removo as linhas 4 primeiras linhas
df_operadores = df_operadores.iloc[4:].reset_index(drop=True)

df_operadores.columns = ['id_operador', 'nome_operador', 'equipe', 'supervisor', 'carga_horária']

Validando as alterações

In [35]:
df_operadores.head(5)

,id_operador,nome_operador,equipe,supervisor,carga_horária
0,1132,Enrico Duarte Nascimento,Platina,Nina Mendes Gonçalves,7h12
1,3119,Isabella Moraes Gonçalves,Diamante,Isabel da Mota da Luz,7h12
2,81711,Mariana Gonçalves Aragão,Ouro,Larissa Nunes Pereira,7h12
3,622,Marcela Novaes Freitas,Bronze,Larissa Nunes Pereira,7h12
4,31011,Anthony Barros Sales,Platina,Isabel da Mota da Luz,7h12


Após isso, irei verificar se há duplicidades na coluna "id_operador"

In [36]:
df_operadores['id_operador'].duplicated().any()

True

Foi identificado que sim, há duplicidades. Vamos descobrir quantos quais são

In [37]:
# Verificando a quantidade
df_operadores['id_operador'].duplicated().sum()

9

In [38]:
# Verificando quais sao as duplicidades
df_operadores[df_operadores['id_operador'].duplicated(keep=False)].sort_values(by='nome_operador', ascending=True)

,id_operador,nome_operador,equipe,supervisor,carga_horária
174,12182,Alícia Moreira Duarte,Diamante,Rebeca Teixeira da Rosa,7h12
191,12182,Alícia Moreira Duarte,Prata,Rebeca Teixeira da Rosa,7h12
58,299,Clara das Neves Barbosa,Diamante,Rebeca Teixeira da Rosa,7h12
149,299,Clara das Neves Barbosa,Platina,Vitor Duarte Porto,7h12
328,10124,Enrico Monteiro Moura,Prata,Nina Mendes Gonçalves,7h12
85,10124,Enrico Monteiro Moura,Prata,Nina Mendes Gonçalves,7h12
30,2113,Enzo Castro Duarte,Diamante,Larissa Nunes Pereira,7h12
372,2113,Enzo Castro Duarte,Ouro,Larissa Nunes Pereira,7h12
393,21414,Fernanda Moraes Caldeira,Prata,Isabel da Mota da Luz,7h12
39,21414,Fernanda Moraes Caldeira,Prata,Larissa Nunes Pereira,7h12


Neste caso me deparo com outra situação em que o ideal seria consultar o responsável pelas bases de dados. Como eu suspeitava, o motivo da duplicidade é devido a uma troca de equipes em que o registro anterior se manteve para o id_operador. Como não há quem eu possa consultar nessa situação de teste, irei utilizar o index de cada linha de registro para determinar qual linha manter. Neste caso, irei manter, para cada id_operador, o index maior (julgando que é o registro mais recente)

In [39]:
df_operadores = df_operadores.sort_index().drop_duplicates(subset='id_operador', keep='last')

Agora vamos validar se as duplicidades foram removidas

In [40]:
# Verificando a quantidade
df_operadores['id_operador'].duplicated().sum()

0

Ótimo, agora as duplicidades foram removidas e temos uma base dimensão de operadores bem estruturada para relacionamentos

Padronizando o nome das colunas em minúsculo e sem pontuações.

Verificando os nomes atuais:

In [41]:
df_operadores.columns

Index(['id_operador', 'nome_operador', 'equipe', 'supervisor',
       'carga_horária'],
      dtype='object')

Removendo as pontuações da coluna carga_horária:

In [42]:
df_operadores = df_operadores.rename(columns={'carga_horária' : 'carga_horaria'})

Baixando a base tratada:

In [43]:
df_operadores.to_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Secundarias/tb_operadores.xlsx", index=False)

## Base telefones

Lendo a base:

In [44]:
df_telefones = pd.read_fwf("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Primarias/base_telefones.txt")

Verificando as colunas:

In [45]:
df_telefones.columns

Index(['DDD|numero|Operadora|CPF'], dtype='object')

Notei que para essa base, os dados estão separados com "|", então irei ler novamente a base com o read_csv e sep="|":

In [46]:
df_telefones = pd.read_csv("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Primarias/base_telefones.txt", sep='|')

Verificando novamente as colunas:

In [47]:
df_telefones.columns

Index(['DDD', 'numero', 'Operadora', 'CPF'], dtype='object')

Agora sim temos as colunas separadas corretamente, vou verificar agora a estrutura da base:

In [48]:
df_telefones.head()

,DDD,numero,Operadora,CPF
0,45,449965605,Claro,7911559246
1,56,691984490,Tim,2357674373
2,73,739742466,Oi,54758802617
3,62,739742466,Claro,7115846083
4,88,691984490,Oi,27641341381


Posteriormente farei uma validação dos telefones desta base com os telefones da base df_clientes e, como na df_clientes não há coluna de DDD, irei criar uma outra coluna concatenando a coluna 'DDD' com a coluna 'numero'.

Porém, primeiramente irei converter ambas para o formato string:

In [49]:
# Convertendo para o formato string a partir da função definida no inicio
to_str(df_telefones, 'numero')
to_str(df_telefones, 'DDD')

df_telefones['telefone'] = df_telefones['DDD'] + df_telefones['numero']

Validando a alteração:

In [50]:
df_telefones.head()

,DDD,numero,Operadora,CPF,telefone
0,45,449965605,Claro,7911559246,45449965605
1,56,691984490,Tim,2357674373,56691984490
2,73,739742466,Oi,54758802617,73739742466
3,62,739742466,Claro,7115846083,62739742466
4,88,691984490,Oi,27641341381,88691984490


Padronizando o nome das colunas em minúsculo e sem pontuações.

Verificando os nomes atuais:

In [51]:
df_telefones.columns

Index(['DDD', 'numero', 'Operadora', 'CPF', 'telefone'], dtype='object')

Alterando os nomes das colunas:

In [52]:
df_telefones.columns = ['ddd', 'numero', 'operadora', 'cpf', 'telefone']

Validando as alterações:

In [53]:
df_clientes.columns

Index(['cpf', 'nome', 'cidade', 'estado', 'cep'], dtype='object')

Verificando se há duplicidades na coluna 'telefone':

In [54]:
df_telefones['telefone'].duplicated().any()

True

Alterando a tipagem da coluna 'cpf' para futuros relacionamentos:

In [55]:
to_str(df_telefones, 'cpf')

Agora vou verificar se todos os valores da coluna 'cpf' possuem a mesma quantidade de caracteres:

In [56]:
df_telefones['qtd'] = df_telefones['cpf'].astype(str).str.len()

In [57]:
df_telefones.sample(10, random_state=9)

,ddd,numero,operadora,cpf,telefone,qtd
1396,86,200167118,Oi,55287841833,86200167118,11
1011,91,192835690,Vivo,30115854,91192835690,8
933,33,786088680,Vivo,625605238,33786088680,9
392,25,47424666,Tim,357532683,2547424666,9
2867,34,701577436,Vivo,24296226631,34701577436,11
2956,67,101727676,Tim,11040082,67101727676,8
2709,84,117766354,Vivo,814466386,84117766354,9
938,66,715850364,Embratel,12075593144,66715850364,11
2571,22,762609307,Vivo,7911559246,22762609307,10
345,13,776109224,Oi,85818650425,13776109224,11


In [58]:
df_telefones['cpf'] = df_telefones['cpf'].astype(str).str.zfill(11)

In [59]:
df_telefones['qtd'] = df_telefones['cpf'].astype(str).str.len()

In [60]:
df_telefones.sample(10, random_state=9)

,ddd,numero,operadora,cpf,telefone,qtd
1396,86,200167118,Oi,55287841833,86200167118,11
1011,91,192835690,Vivo,00030115854,91192835690,11
933,33,786088680,Vivo,00625605238,33786088680,11
392,25,47424666,Tim,00357532683,2547424666,11
2867,34,701577436,Vivo,24296226631,34701577436,11
2956,67,101727676,Tim,00011040082,67101727676,11
2709,84,117766354,Vivo,00814466386,84117766354,11
938,66,715850364,Embratel,12075593144,66715850364,11
2571,22,762609307,Vivo,07911559246,22762609307,11
345,13,776109224,Oi,85818650425,13776109224,11


Baixando a base tratada:

In [61]:
df_telefones.to_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Secundarias/tb_telefones.xlsx", index=False)

## Gerando uma base normalizada

A base normalizada irá ser uma junção das colunas mais relevantes dentre os dataframes disponibilizados e tratados até aqui.

Primeiramente irei definir um novo dataframe que será uma cópia do df_chamadas:

In [62]:
df_normalizada = df_chamadas.copy()

Agora irei verificar a quantidade de linhas e colunas atualmente no dataframe. O importante aqui neste ponto é fazer verificações na quantidade de linhas após cada merge (join) com outros dataframes, para verificar se não há duplicidades sendo incluidas no dataframe e/ou a multiplicação de algumas linhas:

In [63]:
df_normalizada.shape

(9998, 12)

Verificado que há 9998 linhas irei realizar a primeira operação de merge. Irei, a partir da coluna 'id_operador', agregar as colunas 'carga_horaria', 'equipe' e 'supervisor' que estão localizadas no df_operadores:

In [64]:
df_normalizada = df_normalizada.merge(df_operadores[['id_operador', 'carga_horaria', 'equipe', 'supervisor']], on='id_operador', how='left')

Realizado com sucesso, agora vamos verificar se houve incremento de linhas no dataframe:

In [65]:
df_normalizada.shape

(9998, 15)

Ótimo, nenhuma linha foi duplicada, o que é reflexo do bom tratamento de dados que foi feito anteriormente.

Agora irei agregar a coluna 'cpf' do df_telefones, utilizando a coluna 'telefone' entre as duas bases para o relacionamento:

In [66]:
df_normalizada = df_normalizada.merge(df_telefones[['telefone', 'cpf']], on='telefone', how='left')

Novamente, verificando a quantidade de linhas no dataframe após a operação de merge:

In [67]:
df_normalizada.shape

(10993, 16)

Tudo está correndo bem até aqui.

Agora irei agregar as colunas 'nome', 'estado', 'cidade' e 'cep' do dataframe df_clientes, utilizando a coluna 'cpf' para fazer o relacionamento:

In [68]:
df_normalizada = df_normalizada.merge(df_clientes[['cpf', 'nome', 'estado', 'cidade', 'cep']], on='cpf', how='left')

Validando novamente a quantidade de linhas:

In [69]:
df_normalizada.shape

(10993, 20)

Ótimo, as operações de merge foram realizadas com sucesso!

Agora com a tabela normalizada criada, vou verificar como ficou a estrutura do dataframe:

In [70]:
df_normalizada.head()

,data,hora,retido_ura,recebido,atendido,tempo_espera,tempo_atendimento,telefone,id_operador,nome_operador,tabulacao,subtabulacao,carga_horaria,equipe,supervisor,cpf,nome,estado,cidade,cep
0,2025-01-29,21:29:18,0,1,0,00:00:52,00:03:53,79193777494,142012,Luna Ferreira da Rocha,4. Canal de Atendimento,4.2 E-mail,7h12,Platina,Vitor Duarte Porto,47113153603,Beatriz Lima,PR,Colombo,83400-000
1,2025-01-07,00:16:27,0,1,1,00:00:45,00:08:34,27803480622,1752,Levi Cunha da Costa,2. Status do Atendimento,2.4 Em Análise,7h12,Prata,Isabel da Mota da Luz,00146196592,Juliana Silva,PB,João Pessoa,58000-000
2,2025-01-05,06:46:34,1,0,0,00:00:24,00:03:31,23796752326,31913,Helena Costa Ferreira,6. Ação Tomada,6.4 Devolução/Reembolso,7h12,Platina,Rebeca Teixeira da Rosa,00277464522,João Costa,RR,Boa Vista,69300-000
3,2025-01-24,14:57:04,0,1,1,00:01:06,00:07:09,85385553021,11204,João Costela Viana,4. Canal de Atendimento,4.2 E-mail,7h12,Platina,Isabel da Mota da Luz,00619849197,Larissa Martins,PR,Maringá,87000-000
4,2025-01-27,23:55:41,0,1,1,00:01:43,00:03:44,33181611619,2134,Ian Pires Peixoto,5. Grau de Satisfação,5.1 Muito Satisfeito,7h12,Platina,Nina Mendes Gonçalves,00425067894,Juliana Pereira,RS,Porto Alegre,90000-000


Vou ajustar a formatação da coluna de data apenas para facilitar:

In [71]:
df_normalizada['data'] = pd.to_datetime(df_normalizada['data']).dt.strftime('%d/%m/%Y')

Verificando a alteração:

In [72]:
df_normalizada.head()

,data,hora,retido_ura,recebido,atendido,tempo_espera,tempo_atendimento,telefone,id_operador,nome_operador,tabulacao,subtabulacao,carga_horaria,equipe,supervisor,cpf,nome,estado,cidade,cep
0,29/01/2025,21:29:18,0,1,0,00:00:52,00:03:53,79193777494,142012,Luna Ferreira da Rocha,4. Canal de Atendimento,4.2 E-mail,7h12,Platina,Vitor Duarte Porto,47113153603,Beatriz Lima,PR,Colombo,83400-000
1,07/01/2025,00:16:27,0,1,1,00:00:45,00:08:34,27803480622,1752,Levi Cunha da Costa,2. Status do Atendimento,2.4 Em Análise,7h12,Prata,Isabel da Mota da Luz,00146196592,Juliana Silva,PB,João Pessoa,58000-000
2,05/01/2025,06:46:34,1,0,0,00:00:24,00:03:31,23796752326,31913,Helena Costa Ferreira,6. Ação Tomada,6.4 Devolução/Reembolso,7h12,Platina,Rebeca Teixeira da Rosa,00277464522,João Costa,RR,Boa Vista,69300-000
3,24/01/2025,14:57:04,0,1,1,00:01:06,00:07:09,85385553021,11204,João Costela Viana,4. Canal de Atendimento,4.2 E-mail,7h12,Platina,Isabel da Mota da Luz,00619849197,Larissa Martins,PR,Maringá,87000-000
4,27/01/2025,23:55:41,0,1,1,00:01:43,00:03:44,33181611619,2134,Ian Pires Peixoto,5. Grau de Satisfação,5.1 Muito Satisfeito,7h12,Platina,Nina Mendes Gonçalves,00425067894,Juliana Pereira,RS,Porto Alegre,90000-000


Certo, agora vou verificar se existem valores nulos:

In [73]:
df_normalizada.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10993 entries, 0 to 10992
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   data               10993 non-null  object
 1   hora               10993 non-null  object
 2   retido_ura         10993 non-null  int64 
 3   recebido           10993 non-null  int64 
 4   atendido           10993 non-null  int64 
 5   tempo_espera       10993 non-null  object
 6   tempo_atendimento  10993 non-null  object
 7   telefone           10993 non-null  object
 8   id_operador        10993 non-null  object
 9   nome_operador      10993 non-null  object
 10  tabulacao          10993 non-null  object
 11  subtabulacao       10993 non-null  object
 12  carga_horaria      10993 non-null  object
 13  equipe             10993 non-null  object
 14  supervisor         10993 non-null  object
 15  cpf                10922 non-null  object
 16  nome               10922 non-null  objec

É possível verificar a diferença na quantidade de valores nas colunas 'cpf', 'nome', 'estado', 'cidade' e 'cep'. 

** neste ponto já tenho criado o diagrama de relacionamento, o que me auxilia no momento da análise.

Vou verificar os cpf's que estão em branco:

In [74]:
df_normalizada_analise_nulos = df_normalizada[df_normalizada['cpf'].isna()].copy()
df_normalizada_analise_nulos['telefone'].unique()

array(['#N/D'], dtype=object)

In [75]:
df_telefones[(df_telefones['ddd'] == '81') & (df_telefones['numero'] == '8675869')]

,ddd,numero,operadora,cpf,telefone,qtd


Neste ponto cheguei a conclusão de que é necessário realizar um tratamento específico no df_telefones.

Quando a coluna 'numero' tiver apenas 7 caracteres, é necessário incluir um 0 no início, para manter o mesmo padrão do df_chamadas. Porém, para garantir que isso não aconteça tambem caso os telefones possuam também 6 ou 5 caracteres, farei também o mesmo tratamento para incluir os zeros.

Farei isso a seguir:

In [76]:
df_telefones['numero'] = df_telefones['numero'].apply(
    lambda x: f'0{x}' if pd.notna(x) and len(str(x)) == 7 else x
)

In [77]:
df_telefones['numero'] = df_telefones['numero'].apply(
    lambda x: f'00{x}' if pd.notna(x) and len(str(x)) == 6 else x
)

In [78]:
df_telefones['numero'] = df_telefones['numero'].apply(
    lambda x: f'000{x}' if pd.notna(x) and len(str(x)) == 5 else x
)

Como precisarei criar novamente a coluna 'telefone' concatenada, irei dropar a atual e criar uma nova. Estou fazendo isso apenas para ser a ordem cronologica do meu raciocínio, em uma situação real eu voltaria no ponto de tratamentos da df_telefones e corrigiria lá para que isso não fosse necessário:

In [79]:
df_telefones = df_telefones.drop(columns=['telefone'])

# Convertendo para o formato string a partir da função definida no inicio
to_str(df_telefones, 'numero')
to_str(df_telefones, 'ddd')

df_telefones['telefone'] = df_telefones['ddd'] + df_telefones['numero']

Validando o tratamento:

In [80]:
df_telefones[(df_telefones['ddd'] == '84') & (df_telefones['numero'] == '03124605')]

,ddd,numero,operadora,cpf,qtd,telefone


Baixando novamente a base de telefones, agora com esse tratamento a mais:

In [81]:
df_telefones.to_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Secundarias/tb_telefones.xlsx", index=False)

Certo, agora apenas para seguir a minha linha de raciocínio, irei criar uma df_normalizada_2 refazendo todos os merges. Irei fazer dessa forma apenas para manter o registro de minhas descobertas e como avancei nos tratamentos, numa situação real de trabalho eu apenas documentaria isso em um arquivo .docx e corrigiria de forma muito mais simples sem criar um df_normalizada_2.

Como o processo irá se repetir, não irei comentar essas linhas de código novamente:

In [82]:
df_normalizada2 = df_chamadas.copy()

In [83]:
df_normalizada2 = df_normalizada2.merge(df_operadores[['id_operador', 'carga_horaria', 'equipe', 'supervisor']], on='id_operador', how='left')

In [84]:
df_normalizada2 = df_normalizada2.merge(df_telefones[['telefone', 'cpf']], on='telefone', how='left')

In [85]:
df_normalizada2 = df_normalizada2.merge(df_clientes[['cpf', 'nome', 'estado', 'cidade', 'cep']], on='cpf', how='left')

Agora vou verificar se ainda há dados nulos:

In [86]:
df_normalizada2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10977 entries, 0 to 10976
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   data               10977 non-null  datetime64[ns]
 1   hora               10977 non-null  object        
 2   retido_ura         10977 non-null  int64         
 3   recebido           10977 non-null  int64         
 4   atendido           10977 non-null  int64         
 5   tempo_espera       10977 non-null  object        
 6   tempo_atendimento  10977 non-null  object        
 7   telefone           10977 non-null  object        
 8   id_operador        10977 non-null  object        
 9   nome_operador      10977 non-null  object        
 10  tabulacao          10977 non-null  object        
 11  subtabulacao       10977 non-null  object        
 12  carga_horaria      10977 non-null  object        
 13  equipe             10977 non-null  object        
 14  superv

Agora sim, temos a mesma quantidade de valores em todos.

Por fim, irei renomear o nome de algumas colunas apenas para garantir a facilidade na interpretação dos dados:

In [87]:
df_normalizada2 = df_normalizada2.rename(columns={
    'cpf' : 'cpf_cli',
    'nome' : 'nome_cli',
    'estado' : 'estado_cli',
    'cidade' : 'cidade_cli',
    'cep' : 'cep_cli'
})

A seguir irei baixar essa base em formato .xlsx para seguir a análise exploratória em outro arquivo .ipynb.


In [88]:
df_normalizada2.to_excel("C:/Users/Igor/Documents/Teste Analista de MIS/Bases Secundarias/tb_normalizada.xlsx", index=False)